File: lib/arTTY/cache.rb
require "fileutils"
require "json"
require "jsoncfg"
require "minitar"
require "pathname"
require "stringio"
require "typhoeus"
require "zlib"

class ArTTY::Cache < JSONConfig
    extend JSONConfig::Keys

    add_key("art")
    add_key("version")

    def art
        return get_art.keys.sort
    end

    def download_and_extract
        # Temporary working directory
        new = Pathname.new(
            "#{@cachedir}/#{@imagesdir}.new"
        ).expand_path
        FileUtils.rm_rf(new) if (new && new.exist?)

        # Download newest tarball
        request = Typhoeus::Request.new(
            [
                "https://gitlab.com/mjwhitta/arTTY_images/-",
                "archive/master/arTTY_images.tgz"
            ].join("/"),
            timeout: 10
        )
        request.on_headers do |response|
            if (response.code != 200)
                raise ArTTY::Error::FailedToDownload.new
            end
        end
        tgz = StringIO.new(request.run.body)
        raise ArTTY::Error::FailedToDownload.new if (tgz.eof?)

        # Extract new art
        tar = Minitar::Input.new(Zlib::GzipReader.new(tgz))
        tar.each do |entry|
            case entry.name
            when /.+\.json\s*$/
                entry.name.gsub!(/^.*generated\//, "")
                entry.prefix.gsub!(/^.*generated\//, "")
                tar.extract_entry(new, entry)
            end
        end

        # Move new art to final location
        old = Pathname.new("#{@cachedir}/#{@imagesdir}").expand_path
        if (new.exist?)
            FileUtils.rm_rf(old) if (old.exist?)
            FileUtils.mv(new, old)
        end
    rescue Interrupt
        raise
    ensure
        # Cleanup
        FileUtils.rm_rf(new) if (new && new.exist?)
    end
    private :download_and_extract

    def get_file_for(name)
        return get_art[name]["file"]
    end

    def get_height_for(name)
        return get_art[name]["height"]
    end

    def get_width_for(name)
        return get_art[name]["width"]
    end

    def initialize(file = nil)
        file ||= "#{ENV["HOME"]}/.cache/arTTY/art.json"
        @defaults = {
            "art" => Hash.new,
            "version" => ArTTY.current_version
        }

        @cachefile = Pathname.new(file).expand_path
        @cachedir = @cachefile.dirname
        @imagesdir = "arTTY_images"
        super(@cachefile, false)

        refresh if (get_version != ArTTY.current_version)
    end

    def refresh(download = false)
        download_and_extract if (download)

        set_art(Hash.new)
        set_version(ArTTY.current_version)

        [
            "#{@cachedir}/#{@imagesdir}",
            "#{ENV["HOME"]}/.config/arTTY/#{@imagesdir}"
        ].each do |dir|
            Dir["#{dir}/**/*.json"].each do |file|
                img = ArTTY::Art.new(file)
                get_art[img.name] = {
                    "file" => file,
                    "height" => img.height,
                    "width" => img.width
                }
            end
        end

        save
    end
end

File: cache.go
package artty

import (
	"archive/tar"
	"compress/gzip"
	"errors"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"gitlab.com/mjwhitta/jsoncfg"
	"gitlab.com/mjwhitta/pathname"
)

type artCache struct {
	cfg *jsoncfg.JSONCfg
}

// Constructor
func newArtCache() *artCache {
	var c = &artCache{
		cfg: jsoncfg.New(filepath.Join(cacheDir, cacheFile)),
	}

	// Initialize defaults
	c.cfg.SetDefault("art", map[string]interface{}{})
	c.cfg.SetDefault("version", Version)
	c.cfg.SaveDefault()
	c.cfg.Reset()

	// Refresh if newer version detected
	if c.cfg.GetString("version") != Version {
		c.refresh()
	}

	return c
}

func (c *artCache) downloadExtract() error {
	var e error
	var g *gzip.Reader
	var images = "https://gitlab.com/mjwhitta/arTTY_images"
	var res *http.Response
	var t *tar.Reader
	var tgz = "/-/archive/master/arTTY_images.tgz"

	// Download tarball
	if res, e = http.Get(images + tgz); e != nil {
		return e
	}
	defer res.Body.Close()

	// Create gzip reader
	if g, e = gzip.NewReader(res.Body); e != nil {
		return e
	}
	defer g.Close()

	// Create tar reader
	t = tar.NewReader(g)

	// Ensure clean up
	defer os.RemoveAll(filepath.Join(cacheDir, imagesDir+".new"))

	// Extract tar
	if e = c.extract(t); e != nil {
		return e
	}

	// Relocate new JSON files
	return c.organize()
}

func (c *artCache) extract(t *tar.Reader) error {
	var e error
	var h *tar.Header
	var filename string
	var r = regexp.MustCompile(`^arTTY_images.+/generated/(.+)`)

	// Read tarball
	for {
		if h, e = t.Next(); e == io.EOF {
			break
		} else if e != nil {
			return e
		}

		if strings.HasSuffix(h.Name, ".json") {
			// Get filename to write to
			filename = filepath.Join(
				cacheDir,
				imagesDir+".new",
				r.FindStringSubmatch(h.Name)[1],
			)

			if e = c.extractFile(filename, t); e != nil {
				return e
			}
		}
	}

	return nil
}

func (c *artCache) extractFile(filename string, t *tar.Reader) error {
	var dirname string
	var e error
	var f *os.File

	dirname = pathname.Dirname(filename)

	// Ensure directory exists
	if e = os.MkdirAll(dirname, os.ModePerm); e != nil {
		return e
	}

	// Create file
	if f, e = os.Create(filename); e != nil {
		return e
	}

	// Extract file from tarball
	_, e = io.Copy(f, t)
	return e
}

func (c *artCache) getFileOf(name string) string {
	return c.cfg.GetString("art", name, "file")
}

func (c *artCache) getHeightOf(name string) int {
	return c.cfg.GetInt("art", name, "height")
}

func (c *artCache) getWidthOf(name string) int {
	return c.cfg.GetInt("art", name, "width")
}

func (c *artCache) list() []string {
	var keys []string

	for key := range c.cfg.GetMap("art") {
		keys = append(keys, key)
	}

	var less = func(i, j int) bool {
		return (strings.ToLower(keys[i]) < strings.ToLower(keys[j]))
	}

	if !sort.SliceIsSorted(keys, less) {
		sort.SliceStable(keys, less)
	}

	return keys
}

func (c *artCache) organize() error {
	var e error
	var newCache = filepath.Join(cacheDir, imagesDir+".new")
	var oldCache = filepath.Join(cacheDir, imagesDir)

	// Ensure new tarball was extracted
	if !pathname.DoesExist(newCache) {
		return errors.New("Failed to download/extract tarball")
	}

	// Delete old cache
	if e = os.RemoveAll(oldCache); e != nil {
		return e
	}

	// Move new cache to final location
	return os.Rename(newCache, oldCache)
}

func (c *artCache) refresh() {
	var art = map[string]interface{}{}

	var addArt = func(path string, info os.FileInfo, e error) error {
		var img *Art

		if e != nil {
			return e
		}

		if strings.HasSuffix(path, ".json") {
			img = New(path)
			art[img.Name] = map[string]interface{}{
				"file":   path,
				"height": img.Height,
				"width":  img.Width,
			}
		}

		return nil
	}

	// Get all JSON files
	filepath.Walk(filepath.Join(cacheDir, imagesDir), addArt)
	filepath.Walk(filepath.Join(customCacheDir, imagesDir), addArt)

	c.cfg.Set("art", art)
	c.cfg.Set("version", Version)
	c.cfg.Save()
}

func (c *artCache) update() error {
	var e error

	if e = c.downloadExtract(); e != nil {
		return e
	}

	c.refresh()

	return nil
}
